<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Vignette for the hancock package • hancock</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Vignette for the hancock package">
<meta property="og:description" content="hancock">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">hancock</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/hancock.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/concepts.html">Concepts on cell types and signatures</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/kevinrue/hancock/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Vignette for the hancock package</h1>
                        <h4 class="author">Kevin Rue-Albrecht</h4>
            <address class="author_afil">
      Kennedy Institute of Rheumatology, University of Oxford, Headington, Oxford OX3 7FY, UK.<br><a class="author_email" href="mailto:#"></a><a href="mailto:kevin.rue-albrecht@kennedy.ox.ac.uk">kevin.rue-albrecht@kennedy.ox.ac.uk</a>
      </address>
                              <h4 class="author">Second Author</h4>
            <address class="author_afil">
      Second Author’s Affiliation<br><a class="author_email" href="mailto:#"></a><a href="mailto:corresponding@author.com">corresponding@author.com</a>
      </address>
                  
            <h4 class="date">25 May 2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/kevinrue/hancock/blob/master/vignettes/hancock.Rmd"><code>vignettes/hancock.Rmd</code></a></small>
      <div class="hidden name"><code>hancock.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      Example usage of the hancock package demonstrated using example data sets.
    </div>
    
<p><strong>Compiled date</strong>: 2020-05-25</p>
<p><strong>Last edited</strong>: 2018-03-08</p>
<p><strong>License</strong>: MIT + file LICENSE</p>
<div id="overview" class="section level1">
<h1 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h1>
<p>The goal of the <em><a href="https://github.com/kevinrue/hancock">hancock</a></em> package is to provide a collection of methods for learning and applying gene signatures associated with cellular phenotypes and identities. Particular focus is given to single-cell data stored in objects derived from the <em><a href="https://bioconductor.org/packages/3.12/SummarizedExperiment">SummarizedExperiment</a></em> class.</p>
</div>
<div id="getting-started" class="section level1">
<h1 class="hasAnchor">
<a href="#getting-started" class="anchor"></a>Getting started</h1>
<div id="setup" class="section level2">
<h2 class="hasAnchor">
<a href="#setup" class="anchor"></a>Setup</h2>
<p>To run an analysis, the first step is to start R and load the <em><a href="https://github.com/kevinrue/hancock">hancock</a></em> package:</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">hancock</span>)</pre></body></html></div>
</div>
<div id="example-data-set" class="section level2">
<h2 class="hasAnchor">
<a href="#example-data-set" class="anchor"></a>Example data set</h2>
<p>In this example, we use count data for 2,700 peripheral blood mononuclear cells (PBMC) obtained using the <a href="www.10xgenomics.com">10X Genomics</a> platform.</p>
<p>First, we fetch the data as a <em><a href="https://bioconductor.org/packages/3.12/SingleCellExperiment">SingleCellExperiment</a></em> object using the <em><a href="https://bioconductor.org/packages/3.12/TENxPBMCData">TENxPBMCData</a></em> package. The first time that the following code chunk is run, users should expect it to take additional time as it downloads data from the web and caches it on their local machine; subsequent evaluations of the same code chunk should only take a few seconds as the data set is then loaded from the local cache.</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">TENxPBMCData</span>)
<span class="no">tenx_pbmc3k</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TENxPBMCData/man/TENxPBMCData.html">TENxPBMCData</a></span>(<span class="kw">dataset</span><span class="kw">=</span><span class="st">"pbmc3k"</span>)
<span class="no">tenx_pbmc3k</span>
<span class="co">#&gt; class: SingleCellExperiment </span>
<span class="co">#&gt; dim: 32738 2700 </span>
<span class="co">#&gt; metadata(0):</span>
<span class="co">#&gt; assays(1): counts</span>
<span class="co">#&gt; rownames(32738): ENSG00000243485 ENSG00000237613 ... ENSG00000215616</span>
<span class="co">#&gt;   ENSG00000215611</span>
<span class="co">#&gt; rowData names(3): ENSEMBL_ID Symbol_TENx Symbol</span>
<span class="co">#&gt; colnames: NULL</span>
<span class="co">#&gt; colData names(11): Sample Barcode ... Individual Date_published</span>
<span class="co">#&gt; reducedDimNames(0):</span>
<span class="co">#&gt; altExpNames(0):</span></pre></body></html></div>
<p>To enter more rapidly into the subject of learning and applying gene signatures, we provide the cluster assignment of cells produced by the <a href="https://satijalab.org/seurat/pbmc3k_tutorial_1_4.html">Guided Clustering Tutorial</a> of the <em><a href="https://CRAN.R-project.org/package=Seurat">Seurat</a></em> package.</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">tenx_pbmc3k</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"Cell"</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">tenx_pbmc3k</span>)))
<span class="no">ident</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="kw">package</span> <span class="kw">=</span> <span class="st">"hancock"</span>, <span class="st">"extdata"</span>, <span class="st">"pbmc3k.ident.rds"</span>))
<span class="no">tenx_pbmc3k</span> <span class="kw">&lt;-</span> <span class="no">tenx_pbmc3k</span>[, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">ident</span>)]
<span class="no">tenx_pbmc3k</span>$<span class="no">seurat.ident</span> <span class="kw">&lt;-</span> <span class="no">ident</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">tenx_pbmc3k</span>$<span class="no">seurat.ident</span>)
<span class="co">#&gt; </span>
<span class="co">#&gt;    0    1    2    3    4    5    6    7 </span>
<span class="co">#&gt; 1151  479  342  308  157  155   32   14</span></pre></body></html></div>
<p>In addition, we store manually curated cell type annotations in the <code>"seurat.celltype"</code> cell metadata. Those will be used below to learn signatures associated with well characterized cell populations.</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="no">tenx_pbmc3k</span>$<span class="no">seurat.celltype</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span>(<span class="no">tenx_pbmc3k</span>$<span class="no">seurat.ident</span>, <span class="kw">labels</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(
    <span class="st">"CD4 T cells"</span>, <span class="st">"CD14+ Monocytes"</span>, <span class="st">"B cells"</span>, <span class="st">"CD8 T cells"</span>,
    <span class="st">"FCGR3A+ Monocytes"</span>, <span class="st">"NK cells"</span>, <span class="st">"Dendritic Cells"</span>, <span class="st">"Megakaryocytes"</span>
))
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="no">tenx_pbmc3k</span>$<span class="no">seurat.celltype</span>)
<span class="co">#&gt; </span>
<span class="co">#&gt;       CD4 T cells   CD14+ Monocytes           B cells       CD8 T cells </span>
<span class="co">#&gt;              1151               479               342               308 </span>
<span class="co">#&gt; FCGR3A+ Monocytes          NK cells   Dendritic Cells    Megakaryocytes </span>
<span class="co">#&gt;               157               155                32                14</span></pre></body></html></div>
</div>
<div id="learning-signatures" class="section level2">
<h2 class="hasAnchor">
<a href="#learning-signatures" class="anchor"></a>Learning signatures</h2>
<p>In order to find markers that discriminate subsets of cells from each other, learning methods typically require prior clustering information. In <em><a href="https://bioconductor.org/packages/3.12/SingleCellExperiment">SingleCellExperiment</a></em> objects, this information is easily stored as a <code>factor</code> in a column of the <code>colData</code> slot.</p>
<p>For instance, the learning method <code>"PositiveProportionDifference"</code> can be applied to identify markers for a set of cell populations. In particular, this method offers a variety of filters on individual markers (e.g., minimal difference in detection rate between the target cluster and any other cluster), and on the combined set of markers (e.g., minimal proportion of cells in the target cluster where all markers are detected simultaneously).</p>
<p>Here, we use the manually curated cell type labels to find genes markers for each population of cells in the PBMC. Specifically, we require markers to detected (strictly more than 0 counts; <code>assay.type = "counts", threshold = 0</code>) at least 20% more frequently in the target cluster than any other cluster (<code>min.diff = 0.2, diff.method = "min"</code>). Furthermore, we also require the combined set of markers to be codetected in at least 10% of the target cluster (<code>min.prop = 0.1</code>). Lastly, we request the method to return a maximum of 2 markers per signature (<code>n = 2</code>).</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="no">basesets</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/learnSignatures.html">learnSignatures</a></span>(
    <span class="kw">se</span> <span class="kw">=</span> <span class="no">tenx_pbmc3k</span>, <span class="kw">assay.type</span> <span class="kw">=</span> <span class="st">"counts"</span>,
    <span class="kw">method</span> <span class="kw">=</span> <span class="st">"PositiveProportionDifference"</span>, <span class="kw">cluster.col</span> <span class="kw">=</span> <span class="st">"seurat.celltype"</span>,
    <span class="kw">threshold</span> <span class="kw">=</span> <span class="fl">0</span>, <span class="kw">n</span> <span class="kw">=</span> <span class="fl">2</span>, <span class="kw">min.diff</span> <span class="kw">=</span> <span class="fl">0.2</span>, <span class="kw">diff.method</span> <span class="kw">=</span> <span class="st">"min"</span>, <span class="kw">min.prop</span> <span class="kw">=</span> <span class="fl">0.1</span>)
<span class="no">basesets</span>
<span class="co">#&gt; Sets with 16 relations between 16 elements and 8 sets</span>
<span class="co">#&gt;              element             set | ProportionPositive</span>
<span class="co">#&gt;          &lt;character&gt;     &lt;character&gt; |          &lt;numeric&gt;</span>
<span class="co">#&gt;  [1] ENSG00000142546     CD4 T cells |           0.648132</span>
<span class="co">#&gt;  [2] ENSG00000100100     CD4 T cells |           0.417897</span>
<span class="co">#&gt;  [3] ENSG00000143546 CD14+ Monocytes |           0.972860</span>
<span class="co">#&gt;  [4] ENSG00000170458 CD14+ Monocytes |           0.661795</span>
<span class="co">#&gt;  [5] ENSG00000156738         B cells |           0.859649</span>
<span class="co">#&gt;  ...             ...             ... .                ...</span>
<span class="co">#&gt; [12] ENSG00000115523        NK cells |            0.96129</span>
<span class="co">#&gt; [13] ENSG00000179639 Dendritic Cells |            0.84375</span>
<span class="co">#&gt; [14] ENSG00000108561 Dendritic Cells |            0.84375</span>
<span class="co">#&gt; [15] ENSG00000163736  Megakaryocytes |            1.00000</span>
<span class="co">#&gt; [16] ENSG00000163737  Megakaryocytes |            1.00000</span>
<span class="co">#&gt;      minDifferenceProportion</span>
<span class="co">#&gt;                    &lt;numeric&gt;</span>
<span class="co">#&gt;  [1]                0.241882</span>
<span class="co">#&gt;  [2]                0.223092</span>
<span class="co">#&gt;  [3]                0.469675</span>
<span class="co">#&gt;  [4]                0.411795</span>
<span class="co">#&gt;  [5]                0.776847</span>
<span class="co">#&gt;  ...                     ...</span>
<span class="co">#&gt; [12]                0.652849</span>
<span class="co">#&gt; [13]                0.829136</span>
<span class="co">#&gt; [14]                0.438885</span>
<span class="co">#&gt; [15]                0.942675</span>
<span class="co">#&gt; [16]                0.937500</span>
<span class="co">#&gt; -----------</span>
<span class="co">#&gt; elementInfo: IdVector with 1 metadata (ProportionPositive)</span>
<span class="co">#&gt;     setInfo: IdVector with 1 metadata (ProportionPositive)</span></pre></body></html></div>
<p>In <em><a href="https://github.com/kevinrue/hancock">hancock</a></em>, learning methods return <code>Sets</code> objects, defined in the <em><a href="https://github.com/kevinrue/unisets">unisets</a></em> package. This container stores relations between elements (e.g., genes) and sets (e.g., signatures), along with optional metadata associated with each relation. In the next section, we explore the various pieces of information populated by <code>method = "PositiveProportionDifference")</code>.</p>
</div>
<div id="visualise-learning" class="section level2">
<h2 class="hasAnchor">
<a href="#visualise-learning" class="anchor"></a>Visualizing learning outputs</h2>
<div id="relation-metadata" class="section level3">
<h3 class="hasAnchor">
<a href="#relation-metadata" class="anchor"></a>Relation metadata</h3>
<p>Notably, the metadata associated with each relation between a marker (<code>"element"</code>) and the signature (<code>"set"</code>) can be flattened in a <code>data.frame</code> format. Specifically, the <code>"PositiveProportionDifference"</code> describes two pieces of information:</p>
<ul>
<li>
<code>"ProportionPositive"</code>, the proportion of cells with detectable expression of the marker in the associated cluster</li>
<li>
<code>"minDifferenceProportion"</code>, the minimal difference betwween the detection rate in the associated cluster compared with that of each other cluster</li>
</ul>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="kw pkg">knitr</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span>(<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="no">basesets</span>)))</pre></body></html></div>
<table class="table">
<thead><tr class="header">
<th align="left">element</th>
<th align="left">set</th>
<th align="right">ProportionPositive</th>
<th align="right">minDifferenceProportion</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">ENSG00000142546</td>
<td align="left">CD4 T cells</td>
<td align="right">0.6481321</td>
<td align="right">0.2418821</td>
</tr>
<tr class="even">
<td align="left">ENSG00000100100</td>
<td align="left">CD4 T cells</td>
<td align="right">0.4178975</td>
<td align="right">0.2230923</td>
</tr>
<tr class="odd">
<td align="left">ENSG00000143546</td>
<td align="left">CD14+ Monocytes</td>
<td align="right">0.9728601</td>
<td align="right">0.4696754</td>
</tr>
<tr class="even">
<td align="left">ENSG00000170458</td>
<td align="left">CD14+ Monocytes</td>
<td align="right">0.6617954</td>
<td align="right">0.4117954</td>
</tr>
<tr class="odd">
<td align="left">ENSG00000156738</td>
<td align="left">B cells</td>
<td align="right">0.8596491</td>
<td align="right">0.7768466</td>
</tr>
<tr class="even">
<td align="left">ENSG00000105369</td>
<td align="left">B cells</td>
<td align="right">0.9356725</td>
<td align="right">0.7481725</td>
</tr>
</tbody>
</table>
<p>Specifically, we can extract the relationships between markers and clusters and annotate them with gene metadata such as gene symbol, stored in the <code>rowData</code> slot of the <code>tenx_pbmc3k</code> object.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="no">markerTable</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html">merge</a></span>(
    <span class="kw">x</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="no">basesets</span>), <span class="kw">y</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="fu">rowData</span>(<span class="no">tenx_pbmc3k</span>)[, <span class="st">"Symbol"</span>, <span class="kw">drop</span><span class="kw">=</span><span class="fl">FALSE</span>]),
    <span class="kw">by.x</span><span class="kw">=</span><span class="st">"element"</span>, <span class="kw">by.y</span><span class="kw">=</span><span class="st">"row.names"</span>, <span class="kw">sort</span><span class="kw">=</span><span class="fl">FALSE</span>
)
<span class="kw pkg">knitr</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span>(<span class="no">markerTable</span>)</pre></body></html></div>
<table class="table">
<thead><tr class="header">
<th align="left">element</th>
<th align="left">set</th>
<th align="right">ProportionPositive</th>
<th align="right">minDifferenceProportion</th>
<th align="left">Symbol</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">ENSG00000142546</td>
<td align="left">CD4 T cells</td>
<td align="right">0.6481321</td>
<td align="right">0.2418821</td>
<td align="left">NOSIP</td>
</tr>
<tr class="even">
<td align="left">ENSG00000100100</td>
<td align="left">CD4 T cells</td>
<td align="right">0.4178975</td>
<td align="right">0.2230923</td>
<td align="left">PIK3IP1</td>
</tr>
<tr class="odd">
<td align="left">ENSG00000143546</td>
<td align="left">CD14+ Monocytes</td>
<td align="right">0.9728601</td>
<td align="right">0.4696754</td>
<td align="left">S100A8</td>
</tr>
<tr class="even">
<td align="left">ENSG00000170458</td>
<td align="left">CD14+ Monocytes</td>
<td align="right">0.6617954</td>
<td align="right">0.4117954</td>
<td align="left">CD14</td>
</tr>
<tr class="odd">
<td align="left">ENSG00000156738</td>
<td align="left">B cells</td>
<td align="right">0.8596491</td>
<td align="right">0.7768466</td>
<td align="left">MS4A1</td>
</tr>
<tr class="even">
<td align="left">ENSG00000105369</td>
<td align="left">B cells</td>
<td align="right">0.9356725</td>
<td align="right">0.7481725</td>
<td align="left">CD79A</td>
</tr>
<tr class="odd">
<td align="left">ENSG00000113088</td>
<td align="left">CD8 T cells</td>
<td align="right">0.5876623</td>
<td align="right">0.4779849</td>
<td align="left">GZMK</td>
</tr>
<tr class="even">
<td align="left">ENSG00000153563</td>
<td align="left">CD8 T cells</td>
<td align="right">0.5032468</td>
<td align="right">0.3989896</td>
<td align="left">CD8A</td>
</tr>
<tr class="odd">
<td align="left">ENSG00000166927</td>
<td align="left">FCGR3A+ Monocytes</td>
<td align="right">0.8025478</td>
<td align="right">0.5415874</td>
<td align="left">MS4A7</td>
</tr>
<tr class="even">
<td align="left">ENSG00000224397</td>
<td align="left">FCGR3A+ Monocytes</td>
<td align="right">0.8216561</td>
<td align="right">0.5252051</td>
<td align="left">SMIM25</td>
</tr>
<tr class="odd">
<td align="left">ENSG00000100453</td>
<td align="left">NK cells</td>
<td align="right">0.9548387</td>
<td align="right">0.6658777</td>
<td align="left">GZMB</td>
</tr>
<tr class="even">
<td align="left">ENSG00000115523</td>
<td align="left">NK cells</td>
<td align="right">0.9612903</td>
<td align="right">0.6528488</td>
<td align="left">GNLY</td>
</tr>
<tr class="odd">
<td align="left">ENSG00000179639</td>
<td align="left">Dendritic Cells</td>
<td align="right">0.8437500</td>
<td align="right">0.8291362</td>
<td align="left">FCER1A</td>
</tr>
<tr class="even">
<td align="left">ENSG00000108561</td>
<td align="left">Dendritic Cells</td>
<td align="right">0.8437500</td>
<td align="right">0.4388847</td>
<td align="left">C1QBP</td>
</tr>
<tr class="odd">
<td align="left">ENSG00000163736</td>
<td align="left">Megakaryocytes</td>
<td align="right">1.0000000</td>
<td align="right">0.9426752</td>
<td align="left">PPBP</td>
</tr>
<tr class="even">
<td align="left">ENSG00000163737</td>
<td align="left">Megakaryocytes</td>
<td align="right">1.0000000</td>
<td align="right">0.9375000</td>
<td align="left">PF4</td>
</tr>
</tbody>
</table>
</div>
<div id="marker-metadata" class="section level3">
<h3 class="hasAnchor">
<a href="#marker-metadata" class="anchor"></a>Marker metadata</h3>
<p>In addition, metadata associated with each unique marker–irrespective of its specific relationships with individual signature–are stored in the metadata columns of the <code>elementInfo</code> slot. Specifically, the <code>"PositiveProportionDifference"</code> describes <code>"ProportionPositive"</code>, the proportion of cells with detectable expression of the markers across the entire data set.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="fu">mcols</span>(<span class="fu">elementInfo</span>(<span class="no">basesets</span>))
<span class="co">#&gt; DataFrame with 16 rows and 1 column</span>
<span class="co">#&gt;                 ProportionPositive</span>
<span class="co">#&gt;                          &lt;numeric&gt;</span>
<span class="co">#&gt; ENSG00000142546           0.429113</span>
<span class="co">#&gt; ENSG00000100100           0.251706</span>
<span class="co">#&gt; ENSG00000143546           0.276725</span>
<span class="co">#&gt; ENSG00000170458           0.144049</span>
<span class="co">#&gt; ENSG00000156738           0.157316</span>
<span class="co">#&gt; ...                            ...</span>
<span class="co">#&gt; ENSG00000115523          0.1800607</span>
<span class="co">#&gt; ENSG00000179639          0.0208491</span>
<span class="co">#&gt; ENSG00000108561          0.3730099</span>
<span class="co">#&gt; ENSG00000163736          0.0288097</span>
<span class="co">#&gt; ENSG00000163737          0.0155421</span></pre></body></html></div>
<p>Using the gene metadata available in the <code>rowData</code> slot of the <code>tenx_pbmc3k</code> object, we can add the gene symbol associated with each marker to the marker metadata.</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="fu">mcols</span>(<span class="fu">elementInfo</span>(<span class="no">basesets</span>)) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(
    <span class="fu">mcols</span>(<span class="fu">elementInfo</span>(<span class="no">basesets</span>)),
    <span class="fu">rowData</span>(<span class="no">tenx_pbmc3k</span>)[
        <span class="fu">ids</span>(<span class="fu">elementInfo</span>(<span class="no">basesets</span>)),
        <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Symbol"</span>, <span class="st">"ENSEMBL_ID"</span>, <span class="st">"Symbol_TENx"</span>)]
)
<span class="fu">mcols</span>(<span class="fu">elementInfo</span>(<span class="no">basesets</span>))
<span class="co">#&gt; DataFrame with 16 rows and 4 columns</span>
<span class="co">#&gt;                 ProportionPositive      Symbol      ENSEMBL_ID Symbol_TENx</span>
<span class="co">#&gt;                          &lt;numeric&gt; &lt;character&gt;     &lt;character&gt; &lt;character&gt;</span>
<span class="co">#&gt; ENSG00000142546           0.429113       NOSIP ENSG00000142546       NOSIP</span>
<span class="co">#&gt; ENSG00000100100           0.251706     PIK3IP1 ENSG00000100100     PIK3IP1</span>
<span class="co">#&gt; ENSG00000143546           0.276725      S100A8 ENSG00000143546      S100A8</span>
<span class="co">#&gt; ENSG00000170458           0.144049        CD14 ENSG00000170458        CD14</span>
<span class="co">#&gt; ENSG00000156738           0.157316       MS4A1 ENSG00000156738       MS4A1</span>
<span class="co">#&gt; ...                            ...         ...             ...         ...</span>
<span class="co">#&gt; ENSG00000115523          0.1800607        GNLY ENSG00000115523        GNLY</span>
<span class="co">#&gt; ENSG00000179639          0.0208491      FCER1A ENSG00000179639      FCER1A</span>
<span class="co">#&gt; ENSG00000108561          0.3730099       C1QBP ENSG00000108561       C1QBP</span>
<span class="co">#&gt; ENSG00000163736          0.0288097        PPBP ENSG00000163736        PPBP</span>
<span class="co">#&gt; ENSG00000163737          0.0155421         PF4 ENSG00000163737         PF4</span></pre></body></html></div>
</div>
<div id="set-metadata" class="section level3">
<h3 class="hasAnchor">
<a href="#set-metadata" class="anchor"></a>Set metadata</h3>
<p>Similarly, metadata associated with each unique signature–irrespective of its specific relationships with individual markers–are stored in the metadata columns of the <code>setInfo</code> slot. Specifically, the <code>"PositiveProportionDifference"</code> describes <code>"ProportionPositive"</code>, the proportion of cells with detectable expression of all markers associated with each signature across the entire data set.</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="fu">mcols</span>(<span class="fu">setInfo</span>(<span class="no">basesets</span>))
<span class="co">#&gt; DataFrame with 8 rows and 1 column</span>
<span class="co">#&gt;                   ProportionPositive</span>
<span class="co">#&gt;                            &lt;numeric&gt;</span>
<span class="co">#&gt; CD4 T cells                 0.126846</span>
<span class="co">#&gt; CD14+ Monocytes             0.217119</span>
<span class="co">#&gt; B cells                     0.172515</span>
<span class="co">#&gt; CD8 T cells                 0.162338</span>
<span class="co">#&gt; FCGR3A+ Monocytes           0.121019</span>
<span class="co">#&gt; NK cells                    0.161290</span>
<span class="co">#&gt; Dendritic Cells             0.125000</span>
<span class="co">#&gt; Megakaryocytes              0.142857</span></pre></body></html></div>
</div>
</div>
<div id="predict-proportion-positive" class="section level2">
<h2 class="hasAnchor">
<a href="#predict-proportion-positive" class="anchor"></a>Applying signatures to predict labels</h2>
<p>Markers <a href="#learning-signatures">learned previously</a> may then be applied on any data set with compatible gene identifiers. Here, we apply the signatures learned above to the training data set itself, to annotate each cluster with its corresponding signature. In particular, we intentionally use the unsupervised cluster assignment rather instead of the manually curated cell type annotation, to simulate the scenario where users wish to automatically annotate unlabelled populations of cells.</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="no">tenx_pbmc3k.hancock</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/predictSignatures.html">predict</a></span>(
    <span class="no">basesets</span>, <span class="no">tenx_pbmc3k</span>, <span class="kw">assay.type</span> <span class="kw">=</span> <span class="st">"counts"</span>,
    <span class="kw">method</span> <span class="kw">=</span> <span class="st">"ProportionPositive"</span>, <span class="kw">cluster.col</span><span class="kw">=</span><span class="st">"seurat.ident"</span>)
<span class="no">tenx_pbmc3k.hancock</span>
<span class="co">#&gt; class: SingleCellExperiment </span>
<span class="co">#&gt; dim: 32738 2638 </span>
<span class="co">#&gt; metadata(1): hancock</span>
<span class="co">#&gt; assays(1): counts</span>
<span class="co">#&gt; rownames(32738): ENSG00000243485 ENSG00000237613 ... ENSG00000215616</span>
<span class="co">#&gt;   ENSG00000215611</span>
<span class="co">#&gt; rowData names(3): ENSEMBL_ID Symbol_TENx Symbol</span>
<span class="co">#&gt; colnames(2638): Cell1 Cell2 ... Cell2699 Cell2700</span>
<span class="co">#&gt; colData names(14): Sample Barcode ... seurat.celltype hancock</span>
<span class="co">#&gt; reducedDimNames(0):</span>
<span class="co">#&gt; altExpNames(0):</span></pre></body></html></div>
<p>In <em><a href="https://github.com/kevinrue/hancock">hancock</a></em>, the <code>predict</code> function populates:</p>
<ul>
<li>the <code>"hancock"</code> column of <code>colData(tenx_pbmc3k.hancock)</code>
</li>
<li>the <code>"hancock"</code> item of <code>metadata(tenx_pbmc3k.hancock)</code>
</li>
</ul>
<p>In the next section, we explore the various pieces of information populated by <code>method = "ProportionPositive"</code>.</p>
</div>
<div id="visualise-prediction" class="section level2">
<h2 class="hasAnchor">
<a href="#visualise-prediction" class="anchor"></a>Visualizing prediction outputs</h2>
<div id="predicted-cell-label" class="section level3">
<h3 class="hasAnchor">
<a href="#predicted-cell-label" class="anchor"></a>Predicted cell label</h3>
<p>The key output of every prediction method is the cell identity predicted for each cell in the object. All prediction methods store this information in <code>colData(sce)[["hancock"]][["prediction"]]</code>, or <code>sce$hancock$prediction</code>, in short.</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(<span class="fu">colData</span>(<span class="no">tenx_pbmc3k.hancock</span>)<span class="kw">[[</span><span class="st">"hancock"</span>]]))
<span class="co">#&gt;              prediction  </span>
<span class="co">#&gt;  CD4 T cells      :1151  </span>
<span class="co">#&gt;  CD14+ Monocytes  : 479  </span>
<span class="co">#&gt;  B cells          : 342  </span>
<span class="co">#&gt;  CD8 T cells      : 308  </span>
<span class="co">#&gt;  FCGR3A+ Monocytes: 157  </span>
<span class="co">#&gt;  NK cells         : 155  </span>
<span class="co">#&gt;  (Other)          :  46</span></pre></body></html></div>
<p>The <code>metadata</code> slot is used to store some required information:</p>
<ul>
<li>
<code>"GeneSets"</code>, the signatures used in the prediction process</li>
<li>
<code>"method"</code>, the unique identifier of the prediction method</li>
<li>
<code>"packageVersion"</code>, the version of the <em><a href="https://github.com/kevinrue/hancock">hancock</a></em> package that performed the prediction</li>
</ul>
<p>Optional method-specific information may be added by each prediction method. For <code>method="ProportionPositive"</code>, those are:</p>
<ul>
<li>
<code>"ProportionPositiveByCluster"</code>, the proportion of cells in each cluster that express all markers in each signature</li>
<li>
<code>"TopSignatureByCluster"</code>, the most frequently detected signature in each cluster</li>
</ul>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="fu">metadata</span>(<span class="no">tenx_pbmc3k.hancock</span>)<span class="kw">[[</span><span class="st">"hancock"</span>]]
<span class="co">#&gt; $GeneSets</span>
<span class="co">#&gt; Sets with 16 relations between 16 elements and 8 sets</span>
<span class="co">#&gt;              element             set | ProportionPositive</span>
<span class="co">#&gt;          &lt;character&gt;     &lt;character&gt; |          &lt;numeric&gt;</span>
<span class="co">#&gt;  [1] ENSG00000142546     CD4 T cells |           0.648132</span>
<span class="co">#&gt;  [2] ENSG00000100100     CD4 T cells |           0.417897</span>
<span class="co">#&gt;  [3] ENSG00000143546 CD14+ Monocytes |           0.972860</span>
<span class="co">#&gt;  [4] ENSG00000170458 CD14+ Monocytes |           0.661795</span>
<span class="co">#&gt;  [5] ENSG00000156738         B cells |           0.859649</span>
<span class="co">#&gt;  ...             ...             ... .                ...</span>
<span class="co">#&gt; [12] ENSG00000115523        NK cells |            0.96129</span>
<span class="co">#&gt; [13] ENSG00000179639 Dendritic Cells |            0.84375</span>
<span class="co">#&gt; [14] ENSG00000108561 Dendritic Cells |            0.84375</span>
<span class="co">#&gt; [15] ENSG00000163736  Megakaryocytes |            1.00000</span>
<span class="co">#&gt; [16] ENSG00000163737  Megakaryocytes |            1.00000</span>
<span class="co">#&gt;      minDifferenceProportion</span>
<span class="co">#&gt;                    &lt;numeric&gt;</span>
<span class="co">#&gt;  [1]                0.241882</span>
<span class="co">#&gt;  [2]                0.223092</span>
<span class="co">#&gt;  [3]                0.469675</span>
<span class="co">#&gt;  [4]                0.411795</span>
<span class="co">#&gt;  [5]                0.776847</span>
<span class="co">#&gt;  ...                     ...</span>
<span class="co">#&gt; [12]                0.652849</span>
<span class="co">#&gt; [13]                0.829136</span>
<span class="co">#&gt; [14]                0.438885</span>
<span class="co">#&gt; [15]                0.942675</span>
<span class="co">#&gt; [16]                0.937500</span>
<span class="co">#&gt; -----------</span>
<span class="co">#&gt; elementInfo: IdVector with 4 metadata (ProportionPositive, Symbol, ...)</span>
<span class="co">#&gt;     setInfo: IdVector with 1 metadata (ProportionPositive)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $method</span>
<span class="co">#&gt; [1] "ProportionPositive"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $packageVersion</span>
<span class="co">#&gt; [1] '0.99.0'</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $ProportionPositiveByCluster</span>
<span class="co">#&gt;                    cluster</span>
<span class="co">#&gt; signature                     0           1           2           3          4</span>
<span class="co">#&gt;   CD4 T cells       0.284100782 0.025052192 0.014619883 0.077922078 0.05095541</span>
<span class="co">#&gt;   CD14+ Monocytes   0.003475239 0.651356994 0.002923977 0.000000000 0.07643312</span>
<span class="co">#&gt;   B cells           0.003475239 0.000000000 0.809941520 0.003246753 0.00000000</span>
<span class="co">#&gt;   CD8 T cells       0.005212858 0.000000000 0.000000000 0.292207792 0.00000000</span>
<span class="co">#&gt;   FCGR3A+ Monocytes 0.002606429 0.079331942 0.000000000 0.000000000 0.67515924</span>
<span class="co">#&gt;   NK cells          0.003475239 0.010438413 0.002923977 0.129870130 0.00000000</span>
<span class="co">#&gt;   Dendritic Cells   0.006081668 0.004175365 0.002923977 0.006493506 0.00000000</span>
<span class="co">#&gt;   Megakaryocytes    0.000000000 0.004175365 0.000000000 0.006493506 0.01910828</span>
<span class="co">#&gt;                    cluster</span>
<span class="co">#&gt; signature                     5       6          7</span>
<span class="co">#&gt;   CD4 T cells       0.032258065 0.03125 0.00000000</span>
<span class="co">#&gt;   CD14+ Monocytes   0.000000000 0.21875 0.07142857</span>
<span class="co">#&gt;   B cells           0.000000000 0.03125 0.00000000</span>
<span class="co">#&gt;   CD8 T cells       0.006451613 0.03125 0.00000000</span>
<span class="co">#&gt;   FCGR3A+ Monocytes 0.000000000 0.00000 0.07142857</span>
<span class="co">#&gt;   NK cells          0.916129032 0.00000 0.07142857</span>
<span class="co">#&gt;   Dendritic Cells   0.012903226 0.68750 0.00000000</span>
<span class="co">#&gt;   Megakaryocytes    0.000000000 0.00000 1.00000000</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $TopSignatureByCluster</span>
<span class="co">#&gt;                 0                 1                 2                 3 </span>
<span class="co">#&gt;       CD4 T cells   CD14+ Monocytes           B cells       CD8 T cells </span>
<span class="co">#&gt;                 4                 5                 6                 7 </span>
<span class="co">#&gt; FCGR3A+ Monocytes          NK cells   Dendritic Cells    Megakaryocytes </span>
<span class="co">#&gt; 8 Levels: CD4 T cells CD14+ Monocytes B cells ... Megakaryocytes</span></pre></body></html></div>
<p>In particular, <code>"ProportionPositiveByCluster"</code> may be visualized as a heat map using the <code>plotProportionPositive</code> method. This view is useful to examine the specificity of each signature for each cluster.</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="fu"><a href="../reference/predictByProportionPositive.html">plotProportionPositive</a></span>(<span class="no">tenx_pbmc3k.hancock</span>, <span class="kw">cluster_rows</span><span class="kw">=</span><span class="fl">FALSE</span>, <span class="kw">cluster_columns</span><span class="kw">=</span><span class="fl">FALSE</span>)</pre></body></html></div>
<p><img src="hancock_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
</div>
</div>
<div id="renaming-signatures" class="section level2">
<h2 class="hasAnchor">
<a href="#renaming-signatures" class="anchor"></a>Renaming signatures</h2>
<div id="manually" class="section level3">
<h3 class="hasAnchor">
<a href="#manually" class="anchor"></a>Manually</h3>
<p>Renaming a set of signatures is as simple as renaming the identifiers of the <code>setInfo</code> slot that stores the signatures. For instance, here we prefix each signature by a unique integer identifier.</p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="fu">ids</span>(<span class="fu">setInfo</span>(<span class="no">basesets</span>)) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span>(<span class="fu">setInfo</span>(<span class="no">basesets</span>)), <span class="st">". "</span>, <span class="fu">ids</span>(<span class="fu">setInfo</span>(<span class="no">basesets</span>)))
<span class="fu">ids</span>(<span class="fu">setInfo</span>(<span class="no">basesets</span>))
<span class="co">#&gt; [1] "1. CD4 T cells"       "2. CD14+ Monocytes"   "3. B cells"          </span>
<span class="co">#&gt; [4] "4. CD8 T cells"       "5. FCGR3A+ Monocytes" "6. NK cells"         </span>
<span class="co">#&gt; [7] "7. Dendritic Cells"   "8. Megakaryocytes"</span></pre></body></html></div>
</div>
<div id="interactively" class="section level3">
<h3 class="hasAnchor">
<a href="#interactively" class="anchor"></a>Interactively</h3>
<p>In addition, the <em><a href="https://github.com/kevinrue/hancock">hancock</a></em> package includes a lightweight <em><a href="https://CRAN.R-project.org/package=shiny">shiny</a></em> app that offers users the possibility to interactively rename signatures while inspecting their features in a <code>SummarizedExperiment</code> object (e.g., count of cells associated with each signature, layout in reduced dimension).</p>
<p>Specifically, the app requires a set of signatures and a <code>SummarizedExperiment</code> object that was previously annotated with those signatures using the <code>predict</code> function. When closed, the app returns the updated set of signatures.</p>
<p>Furthermore, this <em><a href="https://CRAN.R-project.org/package=shiny">shiny</a></em> app automatically detects the presence of optional dimensionality reduction results in <em><a href="https://bioconductor.org/packages/3.12/SingleCellExperiment">SingleCellExperiment</a></em> objects, allowing inspection and annotation of the gene signatures using that information.</p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">scater</span>)
<span class="no">tenx_pbmc3k</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/logNormCounts.html">logNormCounts</a></span>(<span class="no">tenx_pbmc3k</span>)
<span class="no">tenx_pbmc3k</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runPCA.html">runPCA</a></span>(<span class="no">tenx_pbmc3k</span>)
<span class="no">tenx_pbmc3k</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html">runTSNE</a></span>(<span class="no">tenx_pbmc3k</span>)</pre></body></html></div>
<div class="sourceCode" id="cb17"><html><body><pre class="r"><span class="no">tenx_pbmc3k.hancock</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/predictSignatures.html">predict</a></span>(
    <span class="no">basesets</span>, <span class="no">tenx_pbmc3k</span>, <span class="kw">assay.type</span> <span class="kw">=</span> <span class="st">"counts"</span>, <span class="kw">method</span> <span class="kw">=</span> <span class="st">"ProportionPositive"</span>,
    <span class="kw">cluster.col</span><span class="kw">=</span><span class="st">"seurat.celltype"</span>)
<span class="kw">if</span> (<span class="fu"><a href="https://rdrr.io/r/base/interactive.html">interactive</a></span>()) {
    <span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">shiny</span>)
    <span class="no">basesets</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/shiny/man/runApp.html">runApp</a></span>(<span class="fu"><a href="../reference/shinyLabels.html">shinyLabels</a></span>(<span class="no">basesets</span>, <span class="no">tenx_pbmc3k.hancock</span>))
}
<span class="fu">ids</span>(<span class="fu">setInfo</span>(<span class="no">basesets</span>))
<span class="co">#&gt; [1] "1. CD4 T cells"       "2. CD14+ Monocytes"   "3. B cells"          </span>
<span class="co">#&gt; [4] "4. CD8 T cells"       "5. FCGR3A+ Monocytes" "6. NK cells"         </span>
<span class="co">#&gt; [7] "7. Dendritic Cells"   "8. Megakaryocytes"</span></pre></body></html></div>
<p>As an example of plot available in the app, dimensionality reduction may facilitate the identification of cell populations more similar or related to each other.</p>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="fu"><a href="../reference/predictSignatures.html">reducedDimPrediction</a></span>(<span class="no">tenx_pbmc3k.hancock</span>, <span class="kw">highlight</span> <span class="kw">=</span> <span class="st">"6. NK cells"</span>, <span class="kw">redDimType</span> <span class="kw">=</span> <span class="st">"TSNE"</span>)</pre></body></html></div>
<p><img src="hancock_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
</div>
</div>
</div>
<div id="types-of-signatures" class="section level1">
<h1 class="hasAnchor">
<a href="#types-of-signatures" class="anchor"></a>Types of signatures</h1>
<div id="absolute-markers" class="section level2">
<h2 class="hasAnchor">
<a href="#absolute-markers" class="anchor"></a>Absolute markers</h2>
<p>As described in the accompanying <a href="concepts.html">concepts</a> vignette, absolute markers (also known as “pan markers”) may be defined as genes detected in each cluster, irrespective of their expression in the other clusters.</p>
<p>For instance, the <code>"PositiveProportionDifference"</code> learning method can be used to identify such markers, by setting <code>min.diff=0</code> to annul any comparison between the detection frequency in the target cluster and all other clusters.</p>
<p>As the numbers of genes detected in each cluster may be rather large, it is generally a good idea to restrict markers to be detected in a very high fraction of the corresponding cluster, for instance <code>min.prop = 0.9</code>. In addition, the <code>threshold</code> argument may be used to define a minimal threshold of expression level to consider a marker as “detected” in each cell, and the <code>assay.type</code> argument declares the assay to use (e.g., “counts”, “logcouts”, “TPM”).</p>
<p>This ensures that for each cluster, the combined set of markers is simultaneously detected above 1 transcript per million (TPM) in at least 90% of cells in that cluster.</p>
<div class="sourceCode" id="cb19"><html><body><pre class="r"><span class="no">basesets</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/learnSignatures.html">learnSignatures</a></span>(
    <span class="kw">se</span> <span class="kw">=</span> <span class="no">tenx_pbmc3k</span>, <span class="kw">assay.type</span> <span class="kw">=</span> <span class="st">"counts"</span>,
    <span class="kw">method</span> <span class="kw">=</span> <span class="st">"PositiveProportionDifference"</span>, <span class="kw">cluster.col</span> <span class="kw">=</span> <span class="st">"seurat.celltype"</span>,
    <span class="kw">min.diff</span> <span class="kw">=</span> <span class="fl">0</span>, <span class="kw">min.prop</span> <span class="kw">=</span> <span class="fl">0.9</span>, <span class="kw">threshold</span> <span class="kw">=</span> <span class="fl">1</span>)
<span class="kw pkg">knitr</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="fu">ids</span>(<span class="fu">sets</span>(<span class="no">basesets</span>))), <span class="kw">col.names</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Signature"</span>, <span class="st">"Genes"</span>))</pre></body></html></div>
<table class="table">
<thead><tr class="header">
<th align="left">Signature</th>
<th align="right">Genes</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">B cells</td>
<td align="right">14</td>
</tr>
<tr class="even">
<td align="left">CD14+ Monocytes</td>
<td align="right">10</td>
</tr>
<tr class="odd">
<td align="left">CD4 T cells</td>
<td align="right">7</td>
</tr>
<tr class="even">
<td align="left">CD8 T cells</td>
<td align="right">14</td>
</tr>
<tr class="odd">
<td align="left">Dendritic Cells</td>
<td align="right">110</td>
</tr>
<tr class="even">
<td align="left">FCGR3A+ Monocytes</td>
<td align="right">29</td>
</tr>
<tr class="odd">
<td align="left">Megakaryocytes</td>
<td align="right">14</td>
</tr>
<tr class="even">
<td align="left">NK cells</td>
<td align="right">13</td>
</tr>
</tbody>
</table>
</div>
<div id="relative-markers" class="section level2">
<h2 class="hasAnchor">
<a href="#relative-markers" class="anchor"></a>Relative markers</h2>
<p>As described in the accompanying <a href="concepts.html">concepts</a> vignette, relative markers (also known as “key markers”) may be defined by <em>differential</em> analysis against other cells in the same sample.</p>
<p>For instance, the <code>"PositiveProportionDifference"</code> learning method can be used to identify such markers, by setting <code>min.diff</code> to a value greater than 0, in order to subset candidate markers to those detected at a rate at least 50% higher than the detection rate observed in any other cluster.</p>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="no">basesets</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/learnSignatures.html">learnSignatures</a></span>(
    <span class="kw">se</span> <span class="kw">=</span> <span class="no">tenx_pbmc3k</span>, <span class="kw">assay.type</span> <span class="kw">=</span> <span class="st">"counts"</span>,
    <span class="kw">method</span> <span class="kw">=</span> <span class="st">"PositiveProportionDifference"</span>, <span class="kw">cluster.col</span> <span class="kw">=</span> <span class="st">"seurat.celltype"</span>,
    <span class="kw">min.diff</span> <span class="kw">=</span> <span class="fl">0.5</span>, <span class="kw">diff.method</span> <span class="kw">=</span> <span class="st">"min"</span>)
<span class="kw pkg">knitr</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span>(<span class="fu">ids</span>(<span class="fu">sets</span>(<span class="no">basesets</span>))), <span class="kw">col.names</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"Signature"</span>, <span class="st">"Genes"</span>))</pre></body></html></div>
<table class="table">
<thead><tr class="header">
<th align="left">Signature</th>
<th align="right">Genes</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">B cells</td>
<td align="right">4</td>
</tr>
<tr class="even">
<td align="left">Dendritic Cells</td>
<td align="right">2</td>
</tr>
<tr class="odd">
<td align="left">FCGR3A+ Monocytes</td>
<td align="right">3</td>
</tr>
<tr class="even">
<td align="left">Megakaryocytes</td>
<td align="right">22</td>
</tr>
<tr class="odd">
<td align="left">NK cells</td>
<td align="right">4</td>
</tr>
</tbody>
</table>
<p>Among the learning outputs stored in the <code>Sets</code> metadata information, the relation metadata column <code>"minDifferenceProportion"</code> reflects the <code>min.diff=0.5</code> threshold applied when learning the signatures.</p>
<div class="sourceCode" id="cb21"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="fu">mcols</span>(<span class="fu">relations</span>(<span class="no">basesets</span>))<span class="kw">[[</span><span class="st">"minDifferenceProportion"</span>]])
<span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span>
<span class="co">#&gt;  0.5087  0.5918  0.7482  0.7287  0.8536  0.9427</span></pre></body></html></div>
<p>In addition, information about individual markers may be stored as element metadata accessible using the <code>elementInfo</code> and <code>mcols</code> methods as shown below. For instance, the proportion of cells positive for each marker across the entire data set.</p>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="fu">mcols</span>(<span class="fu">elementInfo</span>(<span class="no">basesets</span>))
<span class="co">#&gt; DataFrame with 35 rows and 1 column</span>
<span class="co">#&gt;                 ProportionPositive</span>
<span class="co">#&gt;                          &lt;numeric&gt;</span>
<span class="co">#&gt; ENSG00000156738          0.1573161</span>
<span class="co">#&gt; ENSG00000105369          0.1580743</span>
<span class="co">#&gt; ENSG00000247982          0.0815011</span>
<span class="co">#&gt; ENSG00000100721          0.1000758</span>
<span class="co">#&gt; ENSG00000166927          0.1178923</span>
<span class="co">#&gt; ...                            ...</span>
<span class="co">#&gt; ENSG00000171611          0.0075815</span>
<span class="co">#&gt; ENSG00000100351          0.1106899</span>
<span class="co">#&gt; ENSG00000166333          0.1129644</span>
<span class="co">#&gt; ENSG00000145335          0.0212282</span>
<span class="co">#&gt; ENSG00000115758          0.1008340</span></pre></body></html></div>
<p>Similarly, information about individual sets may be stored as set metadata accessible using the <code>setInfo</code> and <code>mcols</code> methods as shown below. For instance, the proportion of cells simultaneously positive for all markers in the cluster where this signature was defined.</p>
<div class="sourceCode" id="cb23"><html><body><pre class="r"><span class="fu">mcols</span>(<span class="fu">setInfo</span>(<span class="no">basesets</span>))
<span class="co">#&gt; DataFrame with 8 rows and 1 column</span>
<span class="co">#&gt;                   ProportionPositive</span>
<span class="co">#&gt;                            &lt;numeric&gt;</span>
<span class="co">#&gt; CD4 T cells                 1.000000</span>
<span class="co">#&gt; CD14+ Monocytes             1.000000</span>
<span class="co">#&gt; B cells                     0.304094</span>
<span class="co">#&gt; CD8 T cells                 1.000000</span>
<span class="co">#&gt; FCGR3A+ Monocytes           0.426752</span>
<span class="co">#&gt; NK cells                    0.612903</span>
<span class="co">#&gt; Dendritic Cells             0.687500</span>
<span class="co">#&gt; Megakaryocytes              0.142857</span></pre></body></html></div>
<p>For instance, future methods to identify absolute markers could include differential expression between the target cluster and all other clusters to identify candidate markers significantly differentially expressed between clusters.</p>
</div>
</div>
<div id="additional-information" class="section level1">
<h1 class="hasAnchor">
<a href="#additional-information" class="anchor"></a>Additional information</h1>
<p>Bug reports can be posted as issues in the <em><a href="https://github.com/kevinrue/hancock">hancock</a></em> GitHub repository. The GitHub repository is the primary source for development versions of the package, where new functionality is added over time. The authors appreciate well-considered suggestions for improvements or new features, or even better, pull requests.</p>
<p>If you use <em><a href="https://github.com/kevinrue/hancock">hancock</a></em> for your analysis, please cite it as shown below:</p>
<div class="sourceCode" id="cb24"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/citation.html">citation</a></span>(<span class="st">"hancock"</span>)
<span class="co">#&gt; </span>
<span class="co">#&gt; To cite package 'hancock' in publications use:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;   Kevin Rue-Albrecht, Robert Amezquita and Aaron Lun (2018). hancock:</span>
<span class="co">#&gt;   Learn and Apply Cell Type Signatures.</span>
<span class="co">#&gt;   https://github.com/kevinrue/hancock,</span>
<span class="co">#&gt;   http://kevinrue.github.io/hancock.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; A BibTeX entry for LaTeX users is</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;   @Manual{,</span>
<span class="co">#&gt;     title = {hancock: Learn and Apply Cell Type Signatures},</span>
<span class="co">#&gt;     author = {Kevin Rue-Albrecht and Robert Amezquita and Aaron Lun},</span>
<span class="co">#&gt;     year = {2018},</span>
<span class="co">#&gt;     note = {https://github.com/kevinrue/hancock,</span>
<span class="co">#&gt; http://kevinrue.github.io/hancock},</span>
<span class="co">#&gt;   }</span></pre></body></html></div>
</div>
<div id="session-info" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Info</h1>
<div class="sourceCode" id="cb25"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()
<span class="co">#&gt; R Under development (unstable) (2020-05-25 r78569)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 16.04.6 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /home/travis/R-bin/lib/R/lib/libRblas.so</span>
<span class="co">#&gt; LAPACK: /home/travis/R-bin/lib/R/lib/libRlapack.so</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span>
<span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] parallel  stats4    stats     graphics  grDevices utils     datasets </span>
<span class="co">#&gt; [8] methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] scater_1.17.0               ggplot2_3.3.0              </span>
<span class="co">#&gt;  [3] TENxPBMCData_1.7.0          HDF5Array_1.17.0           </span>
<span class="co">#&gt;  [5] rhdf5_2.33.0                SingleCellExperiment_1.11.1</span>
<span class="co">#&gt;  [7] hancock_0.99.0              unisets_0.99.0             </span>
<span class="co">#&gt;  [9] SummarizedExperiment_1.19.4 DelayedArray_0.15.1        </span>
<span class="co">#&gt; [11] matrixStats_0.56.0          Biobase_2.49.0             </span>
<span class="co">#&gt; [13] GenomicRanges_1.41.1        GenomeInfoDb_1.25.0        </span>
<span class="co">#&gt; [15] IRanges_2.23.5              S4Vectors_0.27.7           </span>
<span class="co">#&gt; [17] BiocGenerics_0.35.2         BiocStyle_2.17.0           </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] backports_1.1.7               circlize_0.4.9               </span>
<span class="co">#&gt;   [3] AnnotationHub_2.21.0          BiocFileCache_1.13.0         </span>
<span class="co">#&gt;   [5] plyr_1.8.6                    GSEABase_1.51.0              </span>
<span class="co">#&gt;   [7] shinydashboard_0.7.1          BiocParallel_1.23.0          </span>
<span class="co">#&gt;   [9] digest_0.6.25                 htmltools_0.4.0              </span>
<span class="co">#&gt;  [11] viridis_0.5.1                 magrittr_1.5                 </span>
<span class="co">#&gt;  [13] memoise_1.1.0                 cluster_2.1.0                </span>
<span class="co">#&gt;  [15] ComplexHeatmap_2.5.3          Biostrings_2.57.0            </span>
<span class="co">#&gt;  [17] annotate_1.67.0               pkgdown_1.5.1                </span>
<span class="co">#&gt;  [19] colorspace_1.4-1              blob_1.2.1                   </span>
<span class="co">#&gt;  [21] rappdirs_0.3.1                xfun_0.14                    </span>
<span class="co">#&gt;  [23] dplyr_0.8.5                   crayon_1.3.4                 </span>
<span class="co">#&gt;  [25] RCurl_1.98-1.2                jsonlite_1.6.1               </span>
<span class="co">#&gt;  [27] graph_1.67.0                  glue_1.4.1                   </span>
<span class="co">#&gt;  [29] gtable_0.3.0                  zlibbioc_1.35.0              </span>
<span class="co">#&gt;  [31] XVector_0.29.1                GetoptLong_0.1.8             </span>
<span class="co">#&gt;  [33] BiocSingular_1.5.0            Rhdf5lib_1.11.0              </span>
<span class="co">#&gt;  [35] shape_1.4.4                   scales_1.1.1                 </span>
<span class="co">#&gt;  [37] DBI_1.1.0                     Rcpp_1.0.4.6                 </span>
<span class="co">#&gt;  [39] viridisLite_0.3.0             xtable_1.8-4                 </span>
<span class="co">#&gt;  [41] clue_0.3-57                   bit_1.1-15.2                 </span>
<span class="co">#&gt;  [43] rsvd_1.0.3                    httr_1.4.1                   </span>
<span class="co">#&gt;  [45] RColorBrewer_1.1-2            ellipsis_0.3.1               </span>
<span class="co">#&gt;  [47] farver_2.0.3                  pkgconfig_2.0.3              </span>
<span class="co">#&gt;  [49] XML_3.99-0.3                  dbplyr_1.4.3                 </span>
<span class="co">#&gt;  [51] labeling_0.3                  tidyselect_1.1.0             </span>
<span class="co">#&gt;  [53] rlang_0.4.6                   reshape2_1.4.4               </span>
<span class="co">#&gt;  [55] later_1.0.0                   AnnotationDbi_1.51.0         </span>
<span class="co">#&gt;  [57] munsell_0.5.0                 BiocVersion_3.12.0           </span>
<span class="co">#&gt;  [59] tools_4.1.0                   RSQLite_2.2.0                </span>
<span class="co">#&gt;  [61] ExperimentHub_1.15.0          rintrojs_0.2.2               </span>
<span class="co">#&gt;  [63] evaluate_0.14                 stringr_1.4.0                </span>
<span class="co">#&gt;  [65] fastmap_1.0.1                 yaml_2.2.1                   </span>
<span class="co">#&gt;  [67] knitr_1.28                    bit64_0.9-7                  </span>
<span class="co">#&gt;  [69] fs_1.4.1                      purrr_0.3.4                  </span>
<span class="co">#&gt;  [71] mime_0.9                      compiler_4.1.0               </span>
<span class="co">#&gt;  [73] beeswarm_0.2.3                curl_4.3                     </span>
<span class="co">#&gt;  [75] png_0.1-7                     interactiveDisplayBase_1.27.2</span>
<span class="co">#&gt;  [77] tibble_3.0.1                  stringi_1.4.6                </span>
<span class="co">#&gt;  [79] highr_0.8                     desc_1.2.0                   </span>
<span class="co">#&gt;  [81] lattice_0.20-41               Matrix_1.2-18                </span>
<span class="co">#&gt;  [83] vctrs_0.3.0                   pillar_1.4.4                 </span>
<span class="co">#&gt;  [85] lifecycle_0.2.0               BiocManager_1.30.10          </span>
<span class="co">#&gt;  [87] GlobalOptions_0.1.1           BiocNeighbors_1.7.0          </span>
<span class="co">#&gt;  [89] cowplot_1.0.0                 bitops_1.0-6                 </span>
<span class="co">#&gt;  [91] irlba_2.3.3                   httpuv_1.5.2                 </span>
<span class="co">#&gt;  [93] rtracklayer_1.49.1            R6_2.4.1                     </span>
<span class="co">#&gt;  [95] bookdown_0.19                 promises_1.1.0               </span>
<span class="co">#&gt;  [97] gridExtra_2.3                 vipor_0.4.5                  </span>
<span class="co">#&gt;  [99] MASS_7.3-51.6                 assertthat_0.2.1             </span>
<span class="co">#&gt; [101] rprojroot_1.3-2               rjson_0.2.20                 </span>
<span class="co">#&gt; [103] withr_2.2.0                   GenomicAlignments_1.25.0     </span>
<span class="co">#&gt; [105] Rsamtools_2.5.0               GenomeInfoDbData_1.2.3       </span>
<span class="co">#&gt; [107] grid_4.1.0                    beachmat_2.5.0               </span>
<span class="co">#&gt; [109] rmarkdown_2.1                 DelayedMatrixStats_1.11.0    </span>
<span class="co">#&gt; [111] Rtsne_0.15                    shiny_1.4.0.2                </span>
<span class="co">#&gt; [113] ggbeeswarm_0.6.0</span>
<span class="co"># devtools::session_info()</span></pre></body></html></div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Kevin Rue-Albrecht, Robert Amezquita, Aaron Lun.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
